Update Ezviz Component

RenierM26 opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
-->
The YAML configuration will be imported with this update and can safely be removed afterwards.

All camera state attributes were removed and replaced by binary_sensors, sensors or switches.     

List of old camera state attributes with their new replacements:

"privacy": switch entity, camera_name.privacy
"audio": switch entity, camera_name.sound
"ir_led": switch entity, camera_name.infrared
"state_led": switch entity, camera_name.light
"follow_move": switch entity, camera_name.tracking
"alarm_notify": camara status, "Recording"
"alarm_sound_mod": sensor, camera_name.alarm_sound_mod
"encrypted": binary_sensor, camera_name.encrypted
"local_ip": sensor, camera_name.local_ip
"detection_sensibility": sensor, camera_name.detection_sensibility

## Proposed change
<!--
  Rewrote the integration based on latest scaffolding.
  Added additional camera support.
  Collaborating with BaQs and updated API.
  Added Sensor entities.
-->
- Add config flow to ezviz.
- Add binary sensor, sensor and switch platforms.
- Bump pyezviz to 0.1.8.7: https://github.com/BaQs/pyEzviz/compare/9811b38d469590f261ffc6c0c8e0da1f7740022d...0.1.8.7


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [X] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [X] New feature (which adds functionality to an existing integration)
- [X] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
ffmpeg:
stream:
camera:
  - platform: ezviz
    username: YOUR_USERNAME
    password: YOUR_PASSWORD
    cameras:
      D12345678:
        username: YOUR_CAMERA_USERNAME
        password: YOUR_CAMERA_PASSWORD
      D12345699:
        username: YOUR_CAMERA_USERNAME
        password: YOUR_CAMERA_PASSWORD

```

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/17364

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [X] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [X] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [X] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [X] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [X] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [X] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [X] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

RenierM26 committed:
Update Ezviz Component

RenierM26 commented:
Had problems with my vscode environment earlier. Had to recreate the repo to rebase.

RenierM26 committed:
Update Ezviz for pylint test

RenierM26 committed:
Update Ezviz component pylint tests

RenierM26 committed:
Update Ezviz component tests

RenierM26 committed:
Update Ezviz Component tests

RenierM26 committed:
Update Ezviz component pylint error

RenierM26 commented:
Hi,

I'm almost certain that the Python 3.9 (group 2) is failing because of another integration's test. 

Any advice?

RenierM26 commented:
Never mind, I see it's my flow test that's failing. I'll have a look.

RenierM26 committed:
Fix ezviz component config flow tests

RenierM26 commented:
;-) All checks passed.

RenierM26 committed:
Update ezviz component

kl4ver commented:
When will this PR get merged? It would be really nice if the Ezviz intergration will be working again.

MartinHjelmare commented on homeassistant/components/ezviz/__init__.py:
```diff
@@ -1 +1,147 @@
+async def async_setup(hass: HomeAssistantType, config: dict) -> bool:
+    """Set up the Ezviz integration."""
+    hass.data.setdefault(DOMAIN, {})
+    conf = config.get(DOMAIN, EZVIZ_SCHEMA({}))
```
H
There's no existing config section under the ezviz domain key. We should import the configuration YAML from the camera platform into a config flow instead.

MartinHjelmare commented on homeassistant/components/ezviz/camera.py:
```diff
@@ -1,197 +1,253 @@
-ENTITIES = "entities"
+MIN_TIME_BETWEEN_SESSION_RENEW = timedelta(seconds=90)
 
-CAMERA_SCHEMA = vol.Schema(
```
Please keep the platform schema intact besides adding the deprecation validator.

RenierM26 replied:
Done

MartinHjelmare commented on homeassistant/components/ezviz/camera.py:
```diff
@@ -1,197 +1,253 @@
 )
 
 
-def setup_platform(hass, config, add_entities, discovery_info=None):
```
Import the config to a config flow here from a coroutine function `async_setup_platform`.

MartinHjelmare commented on homeassistant/components/ezviz/camera.py:
```diff
@@ -1,197 +1,253 @@
-            # if privacy == true, the device closed the lid or did a 180° tilt
-            "privacy": self._privacy,
+            # Camera firmware version
+            "sw_version": self.coordinator.data[self._idx]["version"],
```
Static info should not be part of the state attributes. Those are meant for dynamic info related to the state of the entity.

RenierM26 replied:
Done. Added switch, sensor, binary_sensor platforms.

MartinHjelmare commented on homeassistant/components/ezviz/camera.py:
```diff
@@ -1,197 +1,253 @@
+            # Camera firmware version update available?
+            "upgrade_available": self.coordinator.data[self._idx]["upgrade_available"],
+            # if privacy == true, the device closed the lid or did a 180� tilt
+            "privacy": self.coordinator.data[self._idx]["privacy"],
```
Binary measurements should be binary sensors.

Since this PR will be a breaking change we have the opportunity to correct the integration and have it follow our design standards.

RenierM26 replied:
Done. Added platforms (sensor, binary_sensor and switch)

MartinHjelmare commented on homeassistant/components/ezviz/camera.py:
```diff
@@ -1,197 +1,253 @@
-            "detection_sensibility": self._detection_sensibility,
+            "local_ip": self.coordinator.data[self._idx]["local_ip"],
+            # camera's battery level if battery camera
+            "battery_level": self.coordinator.data[self._idx]["battery_level"],
```
Battery level and similar measurements that aren't controlled by a service should be a sensor entity.

RenierM26 replied:
Done.

MartinHjelmare commented on homeassistant/components/ezviz/config_flow.py:
```diff
@@ -0,0 +1,126 @@
+    )
+
+
+@config_entries.HANDLERS.register(DOMAIN)
```
The decorator is legacy. Please just pass the domain name as domain keyword argument to the class.

RenierM26 replied:
Done.

MartinHjelmare commented on tests/components/ezviz/test_config_flow.py:
```diff
@@ -0,0 +1,127 @@
+"""Test the Ezviz config flow."""
```
Make sure to test the import step too.

RenierM26 replied:
Hi @MartinHjelmare ,

I'm having difficulty in understanding pytest. Do you perhaps have links to some documentation?

MartinHjelmare replied:
Here's the pytest docs:
https://docs.pytest.org/en/stable/contents.html

Here's a blog about testing in Python:
https://realpython.com/python-testing/

Here are are short guidelines for testing integrations:
https://developers.home-assistant.io/docs/development_testing#writing-tests-for-integrations

The scaffold script template has tests for the config flow with best practices.

RenierM26 replied:
Thank you for the info. I Busy learning pytest and will add tests.

RenierM26 replied:
Added tests for config flow import method.

MartinHjelmare reviewed (changes requested):
Overall looks good!

Most of the camera attributes should be separate binary sensor and sensor entities instead. This can be done in a follow up PR. But we shouldn't add any new attributes that are in conflict with our standards in this PR.

RenierM26 committed:
Update Ezviz component

RenierM26 committed:
Add sensor platforms

RenierM26 committed:
issue with requirements file

RenierM26 committed:
Update binary_sensor to include switches

RenierM26 committed:
Updates to Ezviz sensors

RenierM26 committed:
Removed enum private method.

MartinHjelmare commented on homeassistant/components/ezviz/switch.py:
```diff
@@ -0,0 +1,98 @@
+    supported_switches = set(supported_switches)
+
+    for idx, camera in enumerate(coordinator.data):
+        if camera.get("switches"):
+            for switch in camera.get("switches"):
+                if switch in supported_switches:
+                    switch_entities.append(EzvizSwitch(coordinator, idx, switch))
```
```suggestion
        if not camera.get("switches"):
            continue
        for switch in camera["switches"]:
            if switch not in supported_switches:
                continue
            switch_entities.append(EzvizSwitch(coordinator, idx, switch))
```

RenierM26 committed:
Fix switch args

RenierM26 committed:
Update homeassistant/components/ezviz/switch.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

RenierM26 committed:
config flow checks login info

RenierM26 committed:
Config_flow now imports ezviz from camera platform

RenierM26 committed:
Update test

MartinHjelmare commented on homeassistant/components/ezviz/config_flow.py:
```diff
@@ -41,70 +37,98 @@ class EzvizConfigFlow(ConfigFlow, domain=DOMAIN):
+        for entry in self.hass.config_entries.async_entries(DOMAIN):
+            if (
+                entry.data[CONF_USERNAME] == data[CONF_USERNAME]
+                and entry.data[CONF_PASSWORD] == data[CONF_PASSWORD]
```
Should we really allow the same user to have different config entries with different passwords? Normally we just id on username. Then we can use our unique id config entries api:
https://developers.home-assistant.io/docs/config_entries_config_flow_handler#unique-ids

RenierM26 replied:
Updated config_flow with unique id based on username. Thank you for the info!

MartinHjelmare commented on homeassistant/components/ezviz/config_flow.py:
```diff
@@ -41,70 +37,98 @@ class EzvizConfigFlow(ConfigFlow, domain=DOMAIN):
+
+        except Exception:  # pylint: disable=broad-except
+            _LOGGER.exception(
+                "Error importing ezviz platform config: unexpected exception."
```
We don't end logging messages with period.

RenierM26 replied:
Removed period from config_flow logging entries.

RenierM26 committed:
Updated config_flow with unique_id and remove period from logging

RenierM26 committed:
Added two camera services and clarified service descryptions in services.yaml

RenierM26 committed:
Fixed variable name mistake with new service

RenierM26 committed:
Added french integration translation

MartinHjelmare commented on homeassistant/components/ezviz/translations/fr.json:
```diff
@@ -0,0 +1,33 @@
+{
```
Please remove all translations except the default English translation. The translations will be handled by Lokalise.

RenierM26 replied:
Done.

RenierM26 committed:
Config_flow add camera rtsp credentials as seperate entities, with user step and import step

RenierM26 commented:
I have added core.config_entries for the camera rtsp credentials. This will allow the user to add/remove/change credentials for rtsp camera stream via the integration. 

I'm not sure if this is the best way to do this and would appreciate any feedback.

RenierM26 committed:
rerun hassfest after rebase

MartinHjelmare commented on homeassistant/components/ezviz/camera.py:
```diff
@@ -32,166 +50,194 @@
         vol.Required(CONF_PASSWORD): cv.string,
-        vol.Optional(CONF_CAMERAS, default={}): {cv.string: CAMERA_SCHEMA},
+        vol.Optional(CONF_REGION, default=DEFAULT_REGION): cv.string,
+        vol.Optional(ATTR_CAMERAS, default={}): {cv.string: CAMERA_SCHEMA},
```
Don't change this config schema. It's legacy and only used when importing the existing config to the config flow.

RenierM26 replied:
Schema now matches original.

MartinHjelmare commented on homeassistant/components/ezviz/camera.py:
```diff
@@ -32,166 +50,194 @@
-        ezviz_client = EzvizClient(account, password)
-        ezviz_client.login()
-        cameras = ezviz_client.load_cameras()
+        _LOGGER.debug("camera constant: %s", cameras_conf)
```
Don't log the config. It contains credentials.

RenierM26 replied:
Removed debug strings.

MartinHjelmare commented on homeassistant/components/ezviz/camera.py:
```diff
@@ -32,166 +50,194 @@
+
+    platform = entity_platform.current_platform.get()
+
+    platform.async_register_entity_service(
```
Please don't add any new features that aren't strictly needed to retain the existing feature set. New features can be added in follow up PRs. This PR should just migrate the integration to use config entries with as little changes as possible.

RenierM26 replied:
Noted and Thank you for all your patience.

MartinHjelmare replied:
Please remove the custom services for now and move that addition to a follow up PR.

RenierM26 replied:
Removed all custom services.

RenierM26 committed:
Removed region from legacy config schema, removed logging in camera platform setup that could contain credentials, removed unused constant.

RenierM26 committed:
Regenerate requirements

RenierM26 committed:
Fix tests and add config_flow import config test

RenierM26 committed:
Added addition test to config_flow to test successfull camera entity create.

RenierM26 commented:
This pull request should resolve these two issues as well:
#42876
#32981

MartinHjelmare commented on tests/components/ezviz/test_config_flow.py:
```diff
@@ -64,17 +72,51 @@ async def test_user_form_unexpected_exception(hass, ezviz_config_flow):
     )
-    assert result["type"] == RESULT_TYPE_ABORT
-    assert result["reason"] == "single_instance_allowed"
+    assert result["type"] == RESULT_TYPE_FORM
```
Please complete the flow in each test to either an error, abort or create entry result.

RenierM26 replied:
Done. Added a mock request to end method with "create entry".

MartinHjelmare commented on homeassistant/components/ezviz/config_flow.py:
```diff
@@ -0,0 +1,218 @@
+        self._abort_if_unique_id_configured()
+
+        camera_rstp_creds = {
+            ATTR_SERIAL: data[0],
```
Please use `CONF_*` constants for all config keys.

MartinHjelmare replied:
Maybe we don't need to store the serial in the data but can just use the entry.unique_id attribute?

RenierM26 replied:
I'm passing the data via camera.py, cleaned up config_flow to pass data directly to async_create_entry.

The code is now simpler but I'm not sure if i completely resolved your request? I can't yet find a way to skip storing the camera serial as the init setup routines depend on this. The camera rtsp entities aren't "real" entities so to speak, and are used just for camera rtsp credentials.

MartinHjelmare replied:
We set the serial as unique_id on the config entry in the config flow when configuring the camera(s). So when the camera platform is loaded we can get the serial via `entry.unique_id`.

RenierM26 replied:
The cameras are loaded from the cloud (IP addresses and what cameras are registered). The camera entries are just used for the local rtsp stream to be able to see a camera feed. 

The camera entries also have a different ids than that of the ezviz account. I'm passing the rtsp username and password to the camera platform via hass.data[DOMAIN][entry.unique_id] and then matching them under camera "async_setup_entry"

If i save the camera rtsp credentials under the "ezviz cloud" entry then there is no way the user can add/remove credentials for new cameras. 

I'm struggling to find a better way of doing this. Creating the ezviz cloud entry as a bridge might be an alternative but this might not follow the standards all that well? (Ezviz cloud entry almost like a hue bridge?)

MartinHjelmare commented on homeassistant/components/ezviz/__init__.py:
```diff
@@ -1 +1,120 @@
+
+MIN_TIME_BETWEEN_UPDATES = timedelta(seconds=30)
+
+PLATFORMS = ["camera", "sensor", "binary_sensor", "switch"]
```
Please sort 🔡.

RenierM26 replied:
VSCode's linter only allow met to sort this way. (auto corrects. to this format or 1 line)

PLATFORMS = [
    "camera",
    "sensor",
    "binary_sensor",
    "switch",
]

Hope this is acceptable?

MartinHjelmare replied:
I mean that `"binary_sensor"` should go before `"camera"`.

RenierM26 replied:
Done.

MartinHjelmare commented on homeassistant/components/ezviz/__init__.py:
```diff
@@ -1 +1,120 @@
+    """Set up Ezviz from a config entry."""
+
+    if entry.data.get(ATTR_SERIAL):
+        hass.data[DOMAIN][entry.unique_id] = entry.data
```
I'd store the entry instead since it has all the entry features and data. If the data attribute is reset we can still use the existing entry instance to get access to the new data.

RenierM26 replied:
Done.

RenierM26 committed:
Add to tests method to end in create entry, config_flow cleanup, use entry instead of entry.data

RenierM26 committed:
Removed all services, sorted platforms in init file.

RenierM26 committed:
Changed RTSP logging to debug from warning. (Forgot to change this before commit)

MartinHjelmare commented on homeassistant/components/ezviz/config_flow.py:
```diff
@@ -0,0 +1,210 @@
+
+    async def async_step_user(self, user_input=None):
+        """Handle a flow initiated by the user."""
+        if self._async_current_entries():
```
Are we sure that the existing entry(ies) is the main entry? What if the user deletes the main entry and runs the config flow again?

RenierM26 replied:
Added "type" value under data, check if account type is present before evaluating which form to present the user with.

MartinHjelmare replied:
Remove the check here. Just iterate.

RenierM26 committed:
Cleanup typing, change platform order, bump pyezviz version

RenierM26 committed:
Added types to entries, allow creation of main entry if deleted by validating existance of type

MartinHjelmare commented on homeassistant/components/ezviz/config_flow.py:
```diff
@@ -80,8 +91,15 @@ def async_get_options_flow(config_entry):
+                if item.data.get(CONF_TYPE) == ATTR_TYPE_CLOUD:
+                    return await self.async_step_user_camera()
+
+        self._abort_if_unique_id_configured()
```
We need to set unique id for the config entry before calling this.

MartinHjelmare commented on homeassistant/components/ezviz/__init__.py:
```diff
@@ -1 +1,123 @@
+    """Set up Ezviz from a config entry."""
+
+    if entry.data.get(ATTR_SERIAL):
+        hass.data[DOMAIN][entry.unique_id] = entry
```
We don't really need to save the camera entries in hass.data. The entries are stored in the config entries registry and we have access to them via `hass.config_entries.async_entries`.

We can access that in the camera platform and check for matching entries.

MartinHjelmare commented on homeassistant/components/ezviz/__init__.py:
```diff
@@ -1 +1,123 @@
+async def async_setup_entry(hass, entry):
+    """Set up Ezviz from a config entry."""
+
+    if entry.data.get(ATTR_SERIAL):
```
I'd be more explicit and store a string, eg `CONF_ENTRY_TYPE: "camera"`, in the entry data to represent if it's a camera entry or not.

The serial number is already set as unique_id on the entry so we already have access to that when we need it.

RenierM26 committed:
Config_flow doesn't store serial under entry data, camera rtsp read from entry and not stored in hass, removed duplicate abort if unique id from config flow

MartinHjelmare commented on homeassistant/components/ezviz/config_flow.py:
```diff
@@ -0,0 +1,225 @@
+            {
+                vol.Required(CONF_USERNAME, default=DEFAULT_CAMERA_USERNAME): str,
+                vol.Required(CONF_PASSWORD): str,
+                vol.Required(ATTR_SERIAL): str,
```
I'm assuming it's not possible auto detect the serial number of the camera?

RenierM26 replied:
Correct. I need the serial to know what credentials match what camera.

RenierM26 replied:
Actually you just gave me an idea. I could query the serials from the ezviz cloud account and trigger a multi step flow for the password on the user submitted form.

Is there any way I can trigger a reload of the main integration if someone adds a camera entry via the integration? (It needs to reload in order to rebuild the rtsp url with the correct credentials)

RenierM26 replied:
I've added a discovery flow when a camera is detected without a corresponding entry. The main integration will also reload when a camera entry is added so that the camera rtsp stream make use of the credentials without additional user intervention.

RenierM26 committed:
Fix test of config_flow

MartinHjelmare commented on tests/components/ezviz/test_config_flow.py:
```diff
@@ -95,6 +96,33 @@ async def test_async_step_import_camera(hass, ezviz_config_flow):
+    await async_setup_component(hass, "persistent_notification", {})
+
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": SOURCE_USER}, data=YAML_CONFIG
```
```suggestion
        DOMAIN, context={"source": SOURCE_IMPORT}, data=YAML_CONFIG
```

MartinHjelmare commented on tests/components/ezviz/test_config_flow.py:
```diff
@@ -0,0 +1,161 @@
+    assert result["data"][CONF_FFMPEG_ARGUMENTS] == "/H.264"
+    assert result["data"][CONF_TIMEOUT] == 25
+
+
+class InvalidAuth(exceptions.HomeAssistantError):
+    """Error to indicate there is invalid auth."""
```
```suggestion
```

RenierM26 committed:
Update tests/components/ezviz/test_config_flow.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

RenierM26 committed:
Update tests/components/ezviz/test_config_flow.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

RenierM26 committed:
Update tests/components/ezviz/test_config_flow.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

RenierM26 committed:
Bumped pyezviz api version, added api pyezvizerror exception raised in api (on HTTPError), cleanup unused imports.

RenierM26 committed:
rebase

RenierM26 committed:
cleanup coordinator, bump pyezviz api version, move async_setup_entry to add entry options to camera entries. (order change)

RenierM26 committed:
Added discovery step in config_flow if cameras detected without rtsp config entry

RenierM26 committed:
Reload main integration after addition or completion of camera rtsp config entry

RenierM26 committed:
Add tests for discovery config_flow, added a few other output asserts

MartinHjelmare commented on homeassistant/components/ezviz/camera.py:
```diff
@@ -109,11 +110,23 @@ async def async_setup_entry(hass, entry, async_add_entities):
 
         else:
+
+            await hass.config_entries.flow.async_init(
```
Create a task instead of awaiting. Use `hass.async_create_task`.

RenierM26 replied:
Done.

MartinHjelmare commented on homeassistant/components/ezviz/camera.py:
```diff
@@ -109,11 +110,23 @@ async def async_setup_entry(hass, entry, async_add_entities):
-                "Found camera with serial %s without configuration. Please add an integration instance to see the camera stream",
+            ffmpeg_arguments = DEFAULT_FFMPEG_ARGUMENTS
+            _LOGGER.warning(
+                "Found camera with serial %s without configuration. Please go to integration to complete setup.",
```
We don't end logging messages with period.

RenierM26 replied:
Done.

MartinHjelmare commented on homeassistant/components/ezviz/config_flow.py:
```diff
@@ -159,6 +160,41 @@ async def async_step_user_camera(self, user_input=None):
             step_id="user_camera", data_schema=camera_schema, errors=errors or {}
         )
 
+    async def async_step_discovery(self, discovery_info=None):
```
Discovery info can never be `None`.

RenierM26 replied:
Removed "=None"

MartinHjelmare commented on homeassistant/components/ezviz/config_flow.py:
```diff
@@ -159,6 +160,41 @@ async def async_step_user_camera(self, user_input=None):
+            }
+        )
+
+        if discovery_info is None:
```
This can't happen.

RenierM26 replied:
Done.

MartinHjelmare commented on homeassistant/components/ezviz/config_flow.py:
```diff
@@ -159,6 +160,41 @@ async def async_step_user_camera(self, user_input=None):
+                    },
+                )
+
+        return await self.async_step_discovery()
```
Don't start this step without discovery info.

RenierM26 replied:
This line is no longer needed so I removed it.

MartinHjelmare commented on homeassistant/components/ezviz/config_flow.py:
```diff
@@ -159,6 +160,41 @@ async def async_step_user_camera(self, user_input=None):
+            await self.async_set_unique_id(discovery_info[ATTR_SERIAL])
+            self._abort_if_unique_id_configured()
+
+            if discovery_info.get(CONF_PASSWORD) is not None:
```
There should be a separate confirm step for discovery.

RenierM26 replied:
Not really sure if I resolved this one correctly. I have created a confirm step and moved the create entry to it.

MartinHjelmare commented on homeassistant/components/ezviz/config_flow.py:
```diff
@@ -0,0 +1,261 @@
+        try:
+            await self.hass.async_add_executor_job(client.login)
+
+        except (PyEzvizError, HTTPError) as err:
```
Why is the library not catching the HTTPError and raising a library specific exception? That seems to be a recent change in the library. Preferably the library should only raise library specific exceptions.

RenierM26 replied:
The library (pyezviz) does. Just thought I'll also cater for general connection errors like packet loss when the users tries to add the main account.  A HTTError would then just return unable to authenticate and prompt for credentials and region again.

The config flow also caters for cloud accounts in other regions. If someone enters the incorrect region code it could cause a HTTPError as the host portion of the url is derived from the region code. ("eu" example: apiieu.ezvizlife.com)

MartinHjelmare replied:
It's ok to catch more exceptions but that should be done by the library and then library specific exceptions raised.

RenierM26 replied:
Done. Using library specific exceptions for general connectivity exceptions and PyEzvizError for authentication failures and invalid data returned from api.

MartinHjelmare commented on homeassistant/components/ezviz/config_flow.py:
```diff
@@ -0,0 +1,261 @@
+        await self.async_set_unique_id(data[CONF_USERNAME])
+        self._abort_if_unique_id_configured()
+
+        # constructor does login call
```
Is this comment still true?

RenierM26 replied:
Still true for the Ezviz HA domain. There can be only one Ezviz cloud account linked to a specific email address.

Could this cause issues if a user has a cloud account, in a different domain, linked to their email address? (Blink or canary for example. - if they where also using email address as unique id)

MartinHjelmare replied:
If creating the instance makes I/O we need to schedule the creation on the executor. But it doesn't look like there are any side effects in the init method. 

https://github.com/BaQs/pyEzviz/blob/39e2b0cbf85d246cdd5510e24b64eda5ea9ed48f/pyezviz/client.py#L37-L54

RenierM26 replied:
I'm changing self._session via methods but this isn't really IO, it just sets a cookie. (This is a side effect if I understand the term correctly)

MartinHjelmare replied:
I'm not following. I'm asking if there are any side effects when we instantiate the client class. I can't see any. Please point to where that is if there are side effects at this point.

RenierM26 replied:
There isn't any when initiating the class. Apologies for the misunderstanding.

MartinHjelmare replied:
Good. Then we can remove this comment.

RenierM26 committed:
Camera platform call discover flow with hass.async_create_task. Fixes to config_flow for discovery step

RenierM26 commented on homeassistant/components/ezviz/__init__.py:
```diff
@@ -58,6 +59,14 @@ async def async_setup_entry(hass, entry):
+            for item in hass.config_entries.async_entries():
+                if item.data.get(CONF_TYPE) == ATTR_TYPE_CLOUD:
+                    _LOGGER.info("Reload Ezviz integration with new camera rtsp entry")
+                    await hass.config_entries.async_reload(item.entry_id)
```
I can't see a reason why users would need to add an additional main account. Cameras can be shared between accounts on EZVIZ cloud.

MartinHjelmare commented on homeassistant/components/ezviz/config_flow.py:
```diff
@@ -172,28 +181,23 @@ async def async_step_discovery(self, discovery_info=None):
-        if discovery_info is not None:
-            await self.async_set_unique_id(discovery_info[ATTR_SERIAL])
-            self._abort_if_unique_id_configured()
+        return self.async_show_form(
```
It's the confirm step that should show the form.

RenierM26 replied:
Done.

MartinHjelmare commented on homeassistant/components/ezviz/config_flow.py:
```diff
@@ -160,9 +158,20 @@ async def async_step_user_camera(self, user_input=None):
         errors = {}
+        # This step is called when user goes to integrations,
+        # to complete config for discovered entry.
+        if discovery_info is not None:
```
The discovery step should always be called with discovery info. So we don't need this check.

RenierM26 replied:
Done.

RenierM26 committed:
Fix config_flow discovery, add check to legacy yaml camera platform import, move camera private method to camera import step

RenierM26 committed:
Remove not needed check from config_flow import step.

MartinHjelmare commented on homeassistant/components/ezviz/camera.py:
```diff
@@ -32,166 +34,159 @@
+                    context={"source": SOURCE_DISCOVERY},
+                    data={
+                        ATTR_SERIAL: camera[ATTR_SERIAL],
+                        CONF_USERNAME: None,
```
We don't need to send items that don't have useful values.

RenierM26 replied:
Done.

MartinHjelmare commented on homeassistant/components/ezviz/config_flow.py:
```diff
@@ -0,0 +1,251 @@
+        await self.async_set_unique_id(discovery_info[ATTR_SERIAL])
+        self._abort_if_unique_id_configured()
+
+        return await self.async_step_confirm(discovery_info)
```
We don't need to pass any discovery info to the confirm step. We've set the serial as unique id above. That's enough.

RenierM26 replied:
Done.

MartinHjelmare commented on homeassistant/components/ezviz/config_flow.py:
```diff
@@ -0,0 +1,251 @@
+
+        return await self.async_step_confirm(discovery_info)
+
+    async def async_step_confirm(self, confirm_info):
```
```suggestion
    async def async_step_confirm(self, user_input=None):
```

RenierM26 replied:
Done.

MartinHjelmare commented on homeassistant/components/ezviz/config_flow.py:
```diff
@@ -0,0 +1,251 @@
+        """Confirm and create entry from discovery step."""
+        errors = {}
+
+        if confirm_info is not None:
```
```suggestion
        if user_input is not None:
```

RenierM26 replied:
Done.

MartinHjelmare commented on homeassistant/components/ezviz/config_flow.py:
```diff
@@ -0,0 +1,251 @@
+        errors = {}
+
+        if confirm_info is not None:
+            if confirm_info.get(CONF_PASSWORD) is not None:
```
```suggestion
```

RenierM26 replied:
Done.

MartinHjelmare commented on homeassistant/components/ezviz/config_flow.py:
```diff
@@ -0,0 +1,251 @@
+        if confirm_info is not None:
+            if confirm_info.get(CONF_PASSWORD) is not None:
+                return self.async_create_entry(
+                    title=confirm_info[ATTR_SERIAL],
```
```suggestion
                    title=self.unique_id,
```

RenierM26 replied:
Done.

MartinHjelmare commented on homeassistant/components/ezviz/config_flow.py:
```diff
@@ -0,0 +1,251 @@
+            {
+                vol.Required(CONF_USERNAME, default=DEFAULT_CAMERA_USERNAME): str,
+                vol.Required(CONF_PASSWORD): str,
+                vol.Required(ATTR_SERIAL, default=self.unique_id): str,
```
```suggestion
```

RenierM26 replied:
How would the user know what camera they are adding credentials to? The integration defaults to the integration name.

That's the reason why I left the serial number field.

![image](https://user-images.githubusercontent.com/66512715/111346293-c09a2800-8686-11eb-8876-aefa1a7e859f.png)

MartinHjelmare replied:
We can use title and description place holders with template strings:
https://github.com/home-assistant/core/blob/f3c74948c3a4d2b1048fd980eba9d240db472504/homeassistant/components/deconz/config_flow.py#L209

https://github.com/home-assistant/core/blob/f3c74948c3a4d2b1048fd980eba9d240db472504/homeassistant/components/deconz/strings.json#L3

https://github.com/home-assistant/core/blob/f3c74948c3a4d2b1048fd980eba9d240db472504/homeassistant/components/deconz/config_flow.py#L253

https://github.com/home-assistant/core/blob/f3c74948c3a4d2b1048fd980eba9d240db472504/homeassistant/components/deconz/strings.json#L22

RenierM26 replied:
Added a description_placeholder. Thank You!

MartinHjelmare commented on homeassistant/components/ezviz/config_flow.py:
```diff
@@ -0,0 +1,251 @@
+        return self.async_show_form(
+            step_id="confirm",
+            data_schema=discovered_camera_schema,
+            errors=errors or {},
```
```suggestion
            errors=errors,
```

RenierM26 replied:
Done.

MartinHjelmare commented on homeassistant/components/ezviz/config_flow.py:
```diff
@@ -0,0 +1,251 @@
+
+        if confirm_info is not None:
+            if confirm_info.get(CONF_PASSWORD) is not None:
+                return self.async_create_entry(
```
Can we test the credentials before we create the camera entry?

RenierM26 replied:
It looks like I can test the RTSP creds with ffmpeg.

ffprobe -hide_banner rtsp://admin:<password>@192.168.1.59:554

Would return stream details or "Server returned 401" unauthorized on incorrect password. Do you know of any integrations that I can use as an example?

MartinHjelmare replied:
Example of what? I recommend our scaffold script to generate example for credential validation in the config flow.

RenierM26 replied:
Done. Created module that sends RTSP DESCRIBE messages to camera to verify credentials.

RenierM26 committed:
Cleanup config_flow

RenierM26 committed:
Added config_flow description for discovered camera

RenierM26 committed:
Reordered description in config_flow confim step.

MartinHjelmare commented on homeassistant/components/ezviz/config_flow.py:
```diff
@@ -0,0 +1,250 @@
+
+        await self.async_set_unique_id(discovery_info[ATTR_SERIAL])
+        self._abort_if_unique_id_configured()
+
```
Maybe set the title context here to better distinguish the flow in the GUI?https://github.com/home-assistant/core/blob/314a5518f1b0043c74581970dceac20547b8615b/homeassistant/components/deconz/config_flow.py#L209

RenierM26 replied:
I agree. Added.

RenierM26 committed:
Added serial to flow_step description for discovered camera, readded camera attributes for rtsp stream url (allows user to check RTSP cred), added local ip and firmware upgade available.

RenierM26 commented on homeassistant/components/ezviz/camera.py:
```diff
@@ -181,6 +181,18 @@ def __init__(
+            # RTSP Stream
+            "RTSP stream": self._rtsp_stream,
+        }
+
```
This is only way I can currently think of for a user to confirm the camera RTSP credentials. You can copy/paste the RTSP string in VLC Player (or other network steam viewer) as a way to test externally.

MartinHjelmare replied:
When does the user need to confirm the RTSP credentials? Please explain the use case a bit more.

RenierM26 replied:
This just makes troubleshooting exceptions easier. 

1) Advanced users trying to use different camera channels/streams will also be able to confirm that they configured options correctly etc. (Could also just get this from integration options) 

2) Confirm that you entered the correct password. - A user could also just re-add the integration. 

2) I could rather add the local IP as a sensor if that makes more sense? It's fetched from the EZVIZ api and someone could have a network condition preventing a camera from displaying a feed.

MartinHjelmare replied:
1. Is not a string enough reason to add it to state attributes.
2.The password should be tested by the flow before it's saved in the config entry.
3. Yes we prefer individual sensors over state attributes.

RenierM26 committed:
Bumped pyezviz version and changed region code to region url. (Russia uses a completly different url). PyEzviz adds a Local IP sensor, removed camera entity attributes.

MartinHjelmare commented on homeassistant/components/ezviz/config_flow.py:
```diff
@@ -0,0 +1,252 @@
+            {
+                vol.Required(CONF_USERNAME): str,
+                vol.Required(CONF_PASSWORD): str,
+                vol.Optional(CONF_REGION, default=DEFAULT_REGION): str,
```
We changed to url it seems.

How will the user know the server url?

RenierM26 replied:
It looks like EZVIZ has 3x isolated regions. The default url will work for 95% of the globe. So far only Russia users will need to change the url.

Would it be possible to use a menu with a "custom" option for someone to type their own url?

MartinHjelmare replied:
We could use `vol.In(regions + ["custom"])` and let the `"custom"` choice lead to a separate step where the user can enter a url.

RenierM26 committed:
Add RSTP describe auth check from API to config_flow

MartinHjelmare commented on homeassistant/components/ezviz/config_flow.py:
```diff
@@ -122,20 +146,25 @@ async def async_step_user_camera(self, user_input=None):
+            except InvalidAuth:
+                errors["base"] = "invalid_auth"
+
+            except AbortFlow:
```
Nothing is raising AbortFlow here.

MartinHjelmare commented on homeassistant/components/ezviz/config_flow.py:
```diff
@@ -158,19 +190,28 @@ async def async_step_confirm(self, user_input=None):
+            except InvalidAuth:
+                errors["base"] = "invalid_auth"
+
+            except AbortFlow:
```
As above.

RenierM26 committed:
url as vol.in options in Config_flow

MartinHjelmare commented on homeassistant/components/ezviz/config_flow.py:
```diff
@@ -129,12 +132,45 @@ async def async_step_user(self, user_input=None):
+            {
+                vol.Required(CONF_USERNAME, default=user_input[CONF_USERNAME]): str,
+                vol.Required(CONF_PASSWORD, default=user_input[CONF_PASSWORD]): str,
+                vol.Required(CONF_REGION, default=DEFAULT_REGION): str,
```
Shouldn't this be url instead of region?

RenierM26 replied:
Done. Changed all region references to URL.

MartinHjelmare commented on homeassistant/components/ezviz/config_flow.py:
```diff
@@ -109,6 +111,10 @@ async def async_step_user(self, user_input=None):
         if user_input is not None:
+
+            if user_input[CONF_REGION] == CONF_CUSTOMIZE:
+                return await self.async_step_user_custom_region_url(user_input)
```
Don't pass the input. Store the credentials and ask for the url in the next step. Then validate that before moving on from the next step.

RenierM26 replied:
Done. Storing in self.context["data"]

MartinHjelmare commented on homeassistant/components/ezviz/config_flow.py:
```diff
@@ -56,7 +58,7 @@ async def _validate_and_create_auth(self, data):
             await self.hass.async_add_executor_job(client.login)
 
-        except (PyEzvizError, HTTPError) as err:
+        except (PyEzvizError, HTTPError, ConnectionError) as err:
```
The library should raise library specific errors only.

RenierM26 replied:
Separated Library specific errors from general connection and invalid host/url/ip errors.

RenierM26 replied:
Busy fixing. I see what you mean.

RenierM26 replied:
Exceptions are now raised by library specific errors.

RenierM26 committed:
Config_flow changes to discovery step, added exceptions, fixed tests, added rtsp config validate module mock to test disovery confirm step

RenierM26 committed:
Add test for config_flow step user_camera

RenierM26 committed:
Added tests for abort flow

RenierM26 committed:
Extend tests on custom url flow step

RenierM26 committed:
Fix exceptions in config_flow, fix test for discovery import exception test

RenierM26 committed:
Bump pyezviz api version

RenierM26 committed:
Bump api version, added config_flow function to wake hybernating camera before testing credentials, removed "user camera" step from config flow not needed as cameras are discovered.

MartinHjelmare commented on homeassistant/components/ezviz/config_flow.py:
```diff
@@ -0,0 +1,338 @@
+    def _wake_camera(self, serial):
+        """Wake a hybernating camera."""
+        if self.hass.data.get(DOMAIN):
+            for item in self.hass.data.get(DOMAIN):
```
We can't rely on the integration here. We need to create a new instance of the ezviz api.

RenierM26 replied:
Done. A new instance of the pyezviz api is created and used in config_flow.

RenierM26 committed:
Create pyezviz Api instance for config_flow wake hybernating camera, fixed tests and added fixture to mock method

RenierM26 committed:
Added alarm_control_panel with support to arm/disarm all cameras, fixed camera is available attribute (returns 2 if unavailable, 1 if available)

RenierM26 committed:
Skip ignored entities when setup up camera RTSP stream

MartinHjelmare commented on homeassistant/components/ezviz/alarm_control_panel.py:
```diff
@@ -0,0 +1,85 @@
+"""Support for Ezviz alarm."""
```
Please don't add this platform in this PR. Let's focus on getting this PR merged, ie no new features just complete the minimum needed to have a working flow and tests for that.

RenierM26 replied:
Removed.

RenierM26 committed:
Remove alarm_control_panel, add additional config_flow tests

RenierM26 committed:
Cleanup tests, add tests for discovery_step.

RenierM26 committed:
Add test for config_flow rtsp test step1 exceptions

RenierM26 committed:
Removed redundant except from second step in test RTSP method

RenierM26 committed:
All tests to CREATE or ABORT, added step exception for general HTTP error so user can retry in case of trasient network condition

RenierM26 commented:
I'm done. Please review and let me know if there are any changes required or questions.

RenierM26 committed:
Ammended tests with output checks for step_id, error, data, create entry method calls.

RenierM26 commented on tests/components/ezviz/test_config_flow.py:
```diff
@@ -287,6 +286,74 @@ async def test_import_exception(hass, ezviz_config_flow):
+    assert result["type"] == RESULT_TYPE_FORM
+    assert result["errors"] == {}
+
+    await async_setup_component(hass, "persistent_notification", {})
```
Removed duplicate set persistent notifications from tests.

RenierM26 commented on tests/components/ezviz/test_config_flow.py:
```diff
@@ -287,6 +286,74 @@ async def test_import_exception(hass, ezviz_config_flow):
+        },
+    )
+
+    assert result["type"] == RESULT_TYPE_FORM
```
Added assert for errors, step_id and result.

RenierM26 committed:
bumped ezviz api now rases library exceptions. Config_flow, coordiantor and init raises library exceptions. Updated test sideeffect for library exceptions

RenierM26 committed:
Bump api version, Create mock ezviz cloud account on discovery tests first to allow more complete testing of step.

MartinHjelmare commented on homeassistant/components/ezviz/config_flow.py:
```diff
@@ -0,0 +1,364 @@
+
+        # Get Ezviz cloud credentials from config entry
+        for item in self._async_current_entries():
+            if item.data.get(CONF_TYPE) == ATTR_TYPE_CLOUD:
```
If there's no match here, we'll error below when trying to log in.

RenierM26 committed:
Add abort to rtsp verification method if cloud account was deleted and add tests

MartinHjelmare reviewed (approved):
Looks good!

If we have full test coverage of the config flow we can merge.

MartinHjelmare commented on homeassistant/components/ezviz/camera.py:
```diff
@@ -32,166 +34,166 @@
+                    context={"source": SOURCE_IMPORT},
+                    data={
+                        ATTR_SERIAL: camera[0],
+                        CONF_USERNAME: camera[1][CONF_USERNAME],
```
```suggestion
                        CONF_USERNAME: camera[CONF_USERNAME],
```

MartinHjelmare commented on homeassistant/components/ezviz/camera.py:
```diff
@@ -32,166 +34,166 @@
+                    data={
+                        ATTR_SERIAL: camera[0],
+                        CONF_USERNAME: camera[1][CONF_USERNAME],
+                        CONF_PASSWORD: camera[1][CONF_PASSWORD],
```
```suggestion
                        CONF_PASSWORD: camera[CONF_PASSWORD],
```

MartinHjelmare commented on homeassistant/components/ezviz/camera.py:
```diff
@@ -1,28 +1,30 @@
+from homeassistant.helpers.update_coordinator import CoordinatorEntity
+
+from .const import (
+    ATTR_CAMERAS,
```
```suggestion
    CONF_CAMERAS,
```

MartinHjelmare commented on homeassistant/components/ezviz/const.py:
```diff
@@ -0,0 +1,42 @@
+
+# Configuration
+ATTR_SERIAL = "serial"
+ATTR_CAMERAS = "cameras"
```
```suggestion
CONF_CAMERAS = "cameras"
```

MartinHjelmare commented on tests/components/ezviz/__init__.py:
```diff
@@ -0,0 +1,125 @@
+from unittest.mock import patch
+
+from homeassistant.components.ezviz.const import (
+    ATTR_CAMERAS,
```
```suggestion
    CONF_CAMERAS,
```

MartinHjelmare commented on tests/components/ezviz/__init__.py:
```diff
@@ -0,0 +1,125 @@
+    CONF_USERNAME: "test-username",
+    CONF_PASSWORD: "test-password",
+    CONF_URL: "apiieu.ezvizlife.com",
+    ATTR_CAMERAS: {
```
```suggestion
    CONF_CAMERAS: {
```

MartinHjelmare reviewed (changes requested):
Sorry, found some remaining clean up.

RenierM26 committed:
Update tests/components/ezviz/__init__.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

RenierM26 committed:
Update homeassistant/components/ezviz/const.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

RenierM26 committed:
Update tests/components/ezviz/__init__.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

RenierM26 committed:
Update homeassistant/components/ezviz/camera.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

RenierM26 committed:
Update homeassistant/components/ezviz/camera.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

RenierM26 committed:
Update homeassistant/components/ezviz/camera.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

RenierM26 committed:
Update homeassistant/components/ezviz/camera.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

RenierM26 committed:
Update homeassistant/components/ezviz/camera.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

RenierM26 committed:
Update homeassistant/components/ezviz/camera.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

RenierM26 committed:
Undo config import change to password key for yaml, move hass.data.setdefault to async_setup_entry and remove async_setup

MartinHjelmare commented:
> Note: None of the services in the hassio documentation "Lovelace" example are present in the original integration. There are no custom services in the original integration even though services.yaml file exists.

What are we talking about here? I'm not following.

Please mention that we import the ezviz camera configuration.yaml section to a config entry managed from the GUI and that the config section in the configuration.yaml can be removed after this.

RenierM26 commented:
I think I accidentally removed a marker. The breaking changes section doesn't show. 

I have added what you suggested.

RenierM26 committed:
Fixed tests by removing _patch_async_setup as this was removed from init.

MartinHjelmare commented:
Just three clean ups remaining above

RenierM26 committed:
Update homeassistant/components/ezviz/camera.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

RenierM26 committed:
Update homeassistant/components/ezviz/camera.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

RenierM26 committed:
Update homeassistant/components/ezviz/camera.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

RenierM26 committed:
Changed L67 on camera config to complete suggestion for cleanup

RenierM26 commented:
> Just three clean ups remaining above

Changed L67 as well.  Tested changes locally and they work.

MartinHjelmare commented:
We removed some state attributes from the camera entity, right? We should mention these in the breaking change paragraph and what the alternatives are for the user.

RenierM26 commented:
> We removed some state attributes from the camera entity, right? We should mention these in the breaking change paragraph and what the alternatives are for the user.

Added old sensor names with their alternatives. Hope the format is acceptable?

MartinHjelmare commented:
Great!

MartinHjelmare commented:
Please add a link to a docs PR in this PR description. Then we can merge.

RenierM26 commented:
> Please add a link to a docs PR in this PR description. Then we can merge.

Added a link to docs PR in description as requested.

MartinHjelmare commented:
Thanks!
